### DISCUSSION TITLE
emxq5.x  maximum heap size reache，changing the configuration according to the documentation or online tutorials but does not work

### DISCUSSION THREAD

**Discussion by javadoge** [emqx/emqx#10312] [Q&A]
A message contains the base64 string of the image，

Problem：
Context: maximum heap size reached, Max Heap Size: 6291456, Total Heap Size: 9100325, Kill: true, Error Logger: true, Message Queue Len: 0, GC Info: [{old_heap_block_size,3581853},{heap_block_size,5472277},{mbuf_size,59456},{recent_size,822523},{stack_size,62},{old_heap_size,0},{heap_size,2474138},{bin_vheap_size,205405},{bin_vheap_block_size,220491},{bin_old_vheap_size,0},{bin_old_vheap_block_size,46422}]

Changing the configuration according to the documentation or online tutorials does not work

If must use base64, how should you solve this problem?

**Comment by kjellwinblad** [comment]
The error message indicates that an Erlang process has run out of heap space, as you might have guessed. Max Heap Size: 6291456 means 6291456 words, which is roughly equal to 50 megabytes. This amount might seem substantial, but it could be insufficient if you have messages with a lot of data. I'm currently investigating whether we have a setting to control the max heap size of processes. Could you share any settings you have already tried to change? Links to documentation or online tutorials that you have attempted to follow would also be helpful.

**Reply by javadoge** [reply]
6291456  ÷ 1024 ÷ 1024 ≈ 6 MB  ？？is my understanding correct

my emqx.conf :

mqtt {
  max_packet_size  =  100MB
}

force_shutdown {
  enable = true
  max_heap_size = 32MB
  max_message_queue_len = 1000
}


vm.args : 

+hmax : 100M

**Comment by kjellwinblad** [comment]
@javadoge 
> 6291456 ÷ 1024 ÷ 1024 ≈ 6 MB ？？is my understanding correct

No, because the max heap size is reported in words (one word is 8 bytes on a 64-bit machine).

I believe the relevant setting in your emqx.conf file is max_heap_size. As the current setting is 32MB and it appears to be too low for your use case, you could try setting it to a higher value. You can try setting it to, for example, 512MB to see if the error goes away.

**Reply by javadoge** [reply]
Changed the max heap size to 512M, but it doesn’t work, the error still shows Max Heap Size: 6291456
![1](https://user-images.githubusercontent.com/5305836/229649637-bd33ba25-0944-41d7-9a4c-82e25c7616e3.png)
![2](https://user-images.githubusercontent.com/5305836/229649641-184d16b2-bf61-4388-b0c7-fc247ec9ec97.png)

**Reply by kjellwinblad** [reply]
Okay, that is strange. I will forward this to someone that might know more about this setting. This might be a bug if the setting is ignored. Have you checked the overriding files that are mentioned in the top of /etc/emqx/emqx.conf?

**Reply by kjellwinblad** [reply]
Also can you restart the node. After you change the config file, the node needs to be restarted for the change to take effect.

This issue might be related: https://github.com/emqx/emqx/issues/9867

**Reply by kjellwinblad** [reply]
Another thing that you could check is if you have a zone config that takes precedence: https://www.emqx.io/docs/en/v5.0/admin/cfg.html#broker-zone

**Reply by javadoge** [reply]
thank you~,
without its additional configuration，and the service has also been restarted

**Comment by kjellwinblad** [comment]
Can you open an emqx console (if you start emqx with `emqx console` you will get into the console) and then check what the command `emqx_conf:get([<<"force_shutdown">>]).` outputs?

**Reply by nosovk** [reply]
Is there a way to launch it on running process?

**Reply by savonarola** [reply]
Hello! Yes, you should
* login (via `ssh` or `kubectl exec` or `docker exec`) into the vm/container where emqx node runs
* run `emqx remote_console` command
* run `emqx_conf:get([force_shutdown]).`

**Reply by ankit-akteena** [reply]
i have set the max_heap_size as 128MB in conf file but when i run the above cmd it shows max_heap_size => 16777216 which is 16MB where shoudl i change to increse it , i am on emqx 5.7 opne source @savonarola

**Reply by savonarola** [reply]
The setting was not applied then. How did you apply it?

**Reply by ankit-akteena** [reply]
I have added the below block in emqx.conf file 

force_shutdown {
    enable = true
    max_mailbox_size = 10000
    max_heap_size = "128MB"
}

**Reply by id** [reply]
@ankit-akteena just adding config to a file does not do anything. You need to tell EMQX about it:

```
emqx ctl conf reload
```

**Reply by ankit-akteena** [reply]
@id oh yes, i did restart it after that ...do i need to do a reload still?

**Reply by id** [reply]
No, restart is more than enough.

But you can check the actual config:

```
emqx ctl conf show force_shutdown
```

**Reply by ankit-akteena** [reply]
ubuntu@ip-10-10-147-24:/var/log/emqx$ sudo emqx ctl conf show force_shutdown
force_shutdown {
  enable = true
  max_heap_size = "512MB"
  max_mailbox_size = 10000
}

change to 512 and restarted , above is what i get , is this confrmation that it is set ?

**Reply by id** [reply]
Yes, it is set.
